File: computing/interview/design-patterns/behavioral/state/csharp/StateDemo.cs
using System;

namespace Interview.DesignPatterns.Behavioral.State
{
    public interface IPlayerState
    {
        void Play(AudioPlayer player);
        void Pause(AudioPlayer player);
    }

    public sealed class AudioPlayer
    {
        private IPlayerState _state;
        public string Track { get; private set; } = string.Empty;

        public AudioPlayer()
        {
            _state = new PausedState();
        }

        public void SetState(IPlayerState state)
        {
            _state = state;
        }

        public void SetTrack(string track)
        {
            Track = track;
            Console.WriteLine($"Current track: {Track}");
        }

        public void Play() => _state.Play(this);
        public void Pause() => _state.Pause(this);
    }

    public sealed class PlayingState : IPlayerState
    {
        public void Play(AudioPlayer player)
        {
            Console.WriteLine($"Already playing {player.Track}");
        }

        public void Pause(AudioPlayer player)
        {
            Console.WriteLine("Pausing playback");
            player.SetState(new PausedState());
        }
    }

    public sealed class PausedState : IPlayerState
    {
        public void Play(AudioPlayer player)
        {
            Console.WriteLine($"Resuming playback of {player.Track}");
            player.SetState(new PlayingState());
        }

        public void Pause(AudioPlayer player)
        {
            Console.WriteLine("Already paused");
        }
    }

    public static class Demo
    {
        public static void Main()
        {
            var player = new AudioPlayer();
            player.SetTrack("Track 1");
            player.Play();
            player.Pause();
            player.Play();
        }
    }
}

File: computing/interview/design-patterns/behavioral/state/java/StateDemo.java
package interview.designpatterns.behavioral.state;

interface PlayerState {
    void play(AudioPlayer player);
    void pause(AudioPlayer player);
}

final class AudioPlayer {
    private PlayerState state = new PausedState();
    private String track = "";

    void setState(PlayerState state) {
        this.state = state;
    }

    void setTrack(String track) {
        this.track = track;
        System.out.printf("Current track: %s%n", this.track);
    }

    String track() {
        return track;
    }

    void play() {
        state.play(this);
    }

    void pause() {
        state.pause(this);
    }
}

final class PlayingState implements PlayerState {
    @Override
    public void play(AudioPlayer player) {
        System.out.printf("Already playing %s%n", player.track());
    }

    @Override
    public void pause(AudioPlayer player) {
        System.out.println("Pausing playback");
        player.setState(new PausedState());
    }
}

final class PausedState implements PlayerState {
    @Override
    public void play(AudioPlayer player) {
        System.out.printf("Resuming playback of %s%n", player.track());
        player.setState(new PlayingState());
    }

    @Override
    public void pause(AudioPlayer player) {
        System.out.println("Already paused");
    }
}

public final class StateDemo {
    private StateDemo() {}

    public static void main(String[] args) {
        AudioPlayer player = new AudioPlayer();
        player.setTrack("Track 1");
        player.play();
        player.pause();
        player.play();
    }
}
